DRILL-6094: Created tests for decimal type enhancements

denysord88 opened the pull request.

kkhatua commented:
[~cchang@maprtech.com], could you review the PR from Denys?

cchang738 commented on framework/resources/Functional/decimal_parquet/InternalCastLiteralAsDecimalWhenDecimalsOn.sql:
```diff
@@ -0,0 +1,6 @@
+SET `planner.enable_decimal_data_type` = true;
```
What is the system default for this setting? What is it set in our before_run.sql? What a RESET does? Do you see potential conflict? Thanks.

cchang738 commented on framework/resources/Functional/decimal_parquet/decimal_parquet.json:
```diff
@@ -0,0 +1,53 @@
+{
+  "testId": "filter_pushdown_plan",
```
Here too.

cchang738 commented:
Also, pre-commit automation run failed:

Execution Failures:

/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CaseIfEmptyStrThenNull.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CastLiteralsAsDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CaseIfEmptyStrThenNullAsDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/binaryDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_avro/other_types.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CastStringToDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/roundDecimals.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_avro/decimal_9_0_fixed.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_avro/decimal_9_0_bytes.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/drill3938.q
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CaseEmptyStringCastStringAsDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols3.q

Data Verification Failures:

/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/modify_types1.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols2.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols5.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols6.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols4.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols1.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/modify_types2.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/seq_modify_types1.q

Timeouts:

/root/drillAutomation/framework-custom/framework/resources/Functional/flatten_operators/general/jira1679.q
/root/drillAutomation/framework-custom/framework/resources/Functional/json/json_kvgenflatten/combo/kvgenflattenbug_DRILL-1766.q

------------------------------------------------------------------------

------------------------------------------------------------------------
ITERATION SUMMARY
------------------------------------------------------------------------
Total Tests                  : 10107
Passing Tests                : 10085
Failing Tests                : 22

> Execution Failures         : 12
> Data Verification Failures : 8
> Plan Verification Failures : 0
> Timeout Failures           : 2

Iteration 1 (of 1) duration: 50.54 min

denysord88 commented:
@cchang738 Test plan was reviewed by developers and QA. I created different task to add some test cases dealing with larger dataset and it will be another PR.

Could you please review this PR now? I renamed titles in JSON files and removed tests that changing system params.

cchang738 commented:
@denysord88 Thanks for making the change. But I still see many test failures. Have you addressed these failures? Many seem related to decimal.

Execution Failures:

/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_avro/decimal_9_0_fixed.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CaseIfEmptyStrThenNullAsDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CastStringToDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/drill3938.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols3.q
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CastLiteralsAsDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/roundDecimals.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_avro/decimal_9_0_bytes.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/binaryDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CaseEmptyStringCastStringAsDecimal.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_parquet/CaseIfEmptyStrThenNull.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/decimal_avro/other_types.sql

Data Verification Failures:

/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols4.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/seq_modify_types1.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/modify_types1.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols1.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols5.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/modify_types2.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols2.q
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/change_metadata/par_add_cols6.q

Plan Verification Failures:

/root/drillAutomation/framework-custom/framework/resources/Functional/ctas/ctas_auto_partition/general/plan/drill4449.q

Timeouts:

/root/drillAutomation/framework-custom/framework/resources/Functional/json/json_kvgenflatten/combo/kvgenflattenbug_DRILL-1766.q
/root/drillAutomation/framework-custom/framework/resources/Functional/flatten_operators/general/jira1679.q

denysord88 commented:
I removed ALTER SYSTEM SET command from ddl file.
Also I replaced in queries internal casting empty strings as NULL to the CASE statement.
Test fails in folder Functional/hive/hive_storage/change_metadata/ are failing now at the master of drill-test-framework and will be fixed in another jira.

@cchang738 Could you review this PR with my updated commit please?

cchang738 commented:
@denysord88 Thanks for making the changes. We have a policy that no PR can be merged with failures. So we need to have a clear run for this PR. Are the failed tests in anyway related to your PR?

denysord88 commented:
@cchang738 If you have failed tests only in folder "Functional/hive/hive_storage/change_metadata/", then this PR is correct. That hive tests are failing because they don't support Hive 2.1 and they don't relates to my PR with Decimals.

denysord88 commented:
Trying to run all Functional tests on test cluster to find conflicts of decimal tests with others.

denysord88 commented:
I updated last commit, added secure drill URL with password to gen script framework/resources/Datasources/ctas/decimal/create_decimals_test_data.sh

@cchang738 could you please review this pull request?

cchang738 commented on framework/resources/Datasources/ctas/decimal/create_decimals_test_data.sh:
```diff
@@ -0,0 +1,9 @@
+then
+  ${DRILL_HOME}/bin/sqlline -n ${USERNAME} -u "jdbc:drill:schema=dfs.$schema_name;drillbit=${DRILL_STORAGE_PLUGIN_SERVER}" --run=${DRILL_TEST_DATA_DIR}/Datasources/ctas/decimal/createParquetWithDecimals.ddl
+else
+  ${DRILL_HOME}/bin/sqlline -n ${USERNAME} -p ${PASSWORD} -u "jdbc:drill:schema=dfs.$schema_name;drillbit=${DRILL_STORAGE_PLUGIN_SERVER}" --run=${DRILL_TEST_DATA_DIR}/Datasources/ctas/decimal/createParquetWithDecimals.ddl
```
This looks good to me. Thanks.

cchang738 commented:
@denysord88 Were you able to run pre-commit functional automation? Please provide your test result for review. Thanks.

denysord88 commented:
@cchang738 CV QA test cluster is not fully configured yet. So I can't run pre-commit functional automation tests. I run only tests on my suite and they passed. Could you please run pre-commit functional tests for my branch and if everything is ok merge the changes?

cchang738 commented:
@denysord88 I still hit some error in preparing data:

Fail to run command /root/drillAutomation/framework-custom/framework/resources/Datasources/ctas/decimal/create_decimals_test_data.sh, errMgs:
 1/8          DROP TABLE IF EXISTS dfs.drillTestDir.`CTAS_CSV_PARQUET_DECIMAL9`;
1 row selected (18.1 seconds)
2/8          CREATE TABLE dfs.drillTestDir.`CTAS_CSV_PARQUET_DECIMAL9` 
AS SELECT 
CASE WHEN columns[0] = '' THEN NULL ELSE CAST(columns[0] AS INT) END id_int, 
CASE WHEN columns[1] = '' THEN NULL ELSE CAST(columns[1] AS DECIMAL(9, 0)) END decimal_9_0, 
CASE WHEN columns[2] = '' THEN NULL ELSE CAST(columns[2] AS DECIMAL(9, 1)) END decimal_9_1, 
CASE WHEN columns[3] = '' THEN NULL ELSE CAST(columns[3] AS DECIMAL(9, 4)) END decimal_9_4, 
CASE WHEN columns[4] = '' THEN NULL ELSE CAST(columns[4] AS DECIMAL(9, 4)) END decimal_without_null_9_4, 
CASE WHEN columns[5] = '' THEN NULL ELSE CAST(columns[5] AS DECIMAL(9, 4)) END decimal_with_null_9_4, 
CASE WHEN columns[6] = '' THEN NULL ELSE CAST(columns[6] AS DECIMAL(9, 8)) END decimal_9_8 
FROM dfs.drillTestDir.`ctas/parquet/DRILL-6094/decimal9.csv`;
1 row selected (51.661 seconds)
3/8          DROP TABLE IF EXISTS dfs.drillTestDir.`CTAS_CSV_PARQUET_DECIMAL18`;
1 row selected (0.66 seconds)
4/8          CREATE TABLE dfs.drillTestDir.`CTAS_CSV_PARQUET_DECIMAL18` 
AS SELECT 
CASE WHEN columns[0] = '' THEN NULL ELSE CAST(columns[0] AS INT) END id_int, 
CASE WHEN columns[1] = '' THEN NULL ELSE CAST(columns[1] AS DECIMAL(18, 0)) END decimal_18_0, 
CASE WHEN columns[2] = '' THEN NULL ELSE CAST(columns[2] AS DECIMAL(18, 1)) END decimal_18_1, 
CASE WHEN columns[3] = '' THEN NULL ELSE CAST(columns[3] AS DECIMAL(18, 4)) END decimal_18_4, 
CASE WHEN columns[4] = '' THEN NULL ELSE CAST(columns[4] AS DECIMAL(18, 4)) END decimal_without_null_18_4, 
CASE WHEN columns[5] = '' THEN NULL ELSE CAST(columns[5] AS DECIMAL(18, 4)) END decimal_with_null_18_4, 
CASE WHEN columns[6] = '' THEN NULL ELSE CAST(columns[6] AS DECIMAL(18, 17)) END decimal_18_17 
FROM dfs.drillTestDir.`ctas/parquet/DRILL-6094/decimal18.csv`;
1 row selected (10.279 seconds)
5/8          DROP TABLE IF EXISTS dfs.drillTestDir.`CTAS_CSV_PARQUET_DECIMAL28`;
1 row selected (0.506 seconds)
6/8          CREATE TABLE dfs.drillTestDir.`CTAS_CSV_PARQUET_DECIMAL28` 
AS SELECT 
CASE WHEN columns[0] = '' THEN NULL ELSE CAST(columns[0] AS INT) END id_int, 
CASE WHEN columns[1] = '' THEN NULL ELSE CAST(columns[1] AS DECIMAL(28, 0)) END decimal_28_0, 
CASE WHEN columns[2] = '' THEN NULL ELSE CAST(columns[2] AS DECIMAL(28, 1)) END decimal_28_1, 
CASE WHEN columns[3] = '' THEN NULL ELSE CAST(columns[3] AS DECIMAL(28, 4)) END decimal_28_4, 
CASE WHEN columns[4] = '' THEN NULL ELSE CAST(columns[4] AS DECIMAL(28, 4)) END decimal_without_null_28_4, 
CASE WHEN columns[5] = '' THEN NULL ELSE CAST(columns[5] AS DECIMAL(28, 4)) END decimal_with_null_28_4, 
CASE WHEN columns[6] = '' THEN NULL ELSE CAST(columns[6] AS DECIMAL(28, 27)) END decimal_28_27 
FROM dfs.drillTestDir.`ctas/parquet/DRILL-6094/decimal28.csv`;
[1;31mError: VALIDATION ERROR: From line 10, column 6 to line 10, column 8: Object 'ctas/parquet/DRILL-6094/decimal28.csv' not found within 'dfs.drillTestDir'

denysord88 commented:
I found the reason of that errors. I used folder "/drill/testdata/ctas/parquet" in hadoop fs, but in another script "Datasources/ctas/create_tables.sh" this folder deleting and then creating again. That script delets all my csv files.
I changed folder of my CSV files for CTAS.
@cchang738 Could you please run again pre-commit functional tests? On our CV Test cluster we don't get errors with decimals now.

cchang738 commented:
Pre-commit test runs now. There are some test failures, but does not look like related to this PR.
+1
